### ISSUE TITLE
AssertionError assert self.kind == TypeKind.FUNCTIONPROTO

### ISSUE THREAD

**Issue by Azzeccagarbugli** [jnikula/hawkmoth#14]
Hey guys!

I'm trying to use this extension **to build my documentation** but I got this annoying problem:

```
Extension error:
Could not import extension hawkmoth (exception: cannot import name 'AutodocReporter' from 'sphinx.ext.autodoc' (/usr/lib/python3.7/site-packages/sphinx/ext/autodoc/__init__.py))
make: *** [Makefile:19: html] Error 2
``` 

This happen each time that I try to launch my `make html`.

I tried to Google the mistake and somehow seems something connected with my machine in which I have installed Arch and this is the most evident [proof](https://bugs.archlinux.org/task/62332) that I was able to find.

This is mine configuration:
```
Description:    Manjaro Linux
Release:        18.0.4
Codename:       Illyria
```
I would like to receive your help and support! Thanks a lot!

**Comment by BrunoMSantos** [comment]
Hi @Azzeccagarbugli, is this ssue #6 perhaps?

Maybe you're just using an old version. There's a `python-sphinx-hawkmoth-git` package in AUR by the way that should get you the latest master.

**Comment by Azzeccagarbugli** [comment]
> python-sphinx-hawkmoth-git

Hi @BrunoMSantos, thanks for the help first of all!

I already checked the issue #6 before creating this new one but without any success, and I also tried to install the package `python-sphinx-hawkmoth-git`in **AUR** as you suggested but I got this error:

```
[ ... ]
python-sphinx: /usr/lib/python3.7/site-packages/sphinx/writers/__pycache__/text.cpython-37.pyc è già presente nel filesystem
python-sphinx: /usr/lib/python3.7/site-packages/sphinx/writers/__pycache__/xml.cpython-37.pyc è già presente nel filesystem
python-sphinx: /usr/lib/python3.7/site-packages/sphinx/writers/html.py è già presente nel filesystem
python-sphinx: /usr/lib/python3.7/site-packages/sphinx/writers/html5.py è già presente nel filesystem
python-sphinx: /usr/lib/python3.7/site-packages/sphinx/writers/latex.py è già presente nel filesystem
python-sphinx: /usr/lib/python3.7/site-packages/sphinx/writers/manpage.py è già presente nel filesystem
python-sphinx: /usr/lib/python3.7/site-packages/sphinx/writers/texinfo.py è già presente nel filesystem
python-sphinx: /usr/lib/python3.7/site-packages/sphinx/writers/text.py è già presente nel filesystem
python-sphinx: /usr/lib/python3.7/site-packages/sphinx/writers/xml.py è già presente nel filesystem
Si sono verificati degli errori, nessun pacchetto è stato aggiornato.
==> ERRORE: 'pacman' non è riuscito ad installare le dipendenze mancanti.
:: Errore compilando il/i pacchetto/i python-sphinx-hawkmoth-git
```

I don't know seriously how I should fix this error, and I really would like to use this awesome extension!

**Comment by jnikula** [comment]
I released v0.4 and pushed to pypi. Upgrading from pypi now should fix the issue.

**Comment by Azzeccagarbugli** [comment]
> I released v0.4 and pushed to pypi. Upgrading from pypi now should fix the issue.

Thanks so much for the **really fast** update, I was finally able to launch my `make html` without any problem and I built my documentation successfully with the extension specified in my `config.py` file.

Simply when now when I try to attach an _autodoc_ reference like this `.. c:autodoc:: ./src/main.c` I got the following issue:
```
Exception occurred:
  File "/usr/lib/python3.7/site-packages/clang/cindex.py", line 2316, in is_function_variadic
    assert self.kind == TypeKind.FUNCTIONPROTO
AssertionError
The full traceback has been saved in /tmp/sphinx-err-f4coz9u_.log, if you want to report the issue to the developers.
Please also report this if it was a user error, so that a better error message can be provided next time.
A bug report can be filed in the tracker at <https://github.com/sphinx-doc/sphinx/issues>. Thanks!
make: *** [Makefile:19: html] Error 2
```
Even if in my `config.py` I set up both the extension and the line `cautodoc_root = os.path.abspath('..')`

Do you know how could I fix the generation?

**Comment by jnikula** [comment]
Three ideas to debug further:

- Paste the full traceback referenced in the error message e.g. `/tmp/sphinx-err-f4coz9u_.log` above but note that it's different every time.

- Try the parser from the command line to bypass Sphinx and simplify the setup: `python3 -m hawkmoth path/to/source.c`

- Simplify the source file to find the minimal failing example.

**Comment by jnikula** [comment]
@BrunoMSantos If the traceback points at our call to `is_function_variadic()` then it would suggest we have `cursor.kind == CursorKind.FUNCTION_DECL` with `cursor.type.kind != TypeKind.FUNCTIONPROTO` and I'd be a bit confused by that.

**Comment by Azzeccagarbugli** [comment]
> * Paste the full traceback referenced in the error message e.g. `/tmp/sphinx-err-f4coz9u_.log` above but note that it's different every time.

Yeah sorry if I didn't paste the full traceback, is now online at this [link](https://paste.ofcode.org/Uvy9bKjZNBeLMquCHAkCmv). 

> * Try the parser from the command line to bypass Sphinx and simplify the setup: `python3 -m hawkmoth path/to/source.c`

I also tried to do this but I got the following issue:

```
$ python3 -m hawkmoth src/main.c

Traceback (most recent call last):
  File "/usr/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/usr/lib/python3.7/site-packages/hawkmoth/__main__.py", line 45, in <module>
    main()
  File "/usr/lib/python3.7/site-packages/hawkmoth/__main__.py", line 34, in main
    docs, errors = parse(args.file, compat=args.compat, clang=args.clang)
  File "/usr/lib/python3.7/site-packages/hawkmoth/parser.py", line 293, in parse
    result.extend(_recursive_parse(comments, cursor, 0, compat))
  File "/usr/lib/python3.7/site-packages/hawkmoth/parser.py", line 224, in _recursive_parse
    if cursor.type.is_function_variadic():
  File "/usr/lib/python3.7/site-packages/clang/cindex.py", line 2316, in is_function_variadic
    assert self.kind == TypeKind.FUNCTIONPROTO
AssertionError

```

**Comment by BrunoMSantos** [comment]
> @BrunoMSantos If the traceback points at our call to is_function_variadic() then it would suggest we have cursor.kind == CursorKind.FUNCTION_DECL with cursor.type.kind != TypeKind.FUNCTIONPROTO and I'd be a bit confused by that.

Hm... Clang's AST can look a bit odd in places to my untrained eyes. Not sure I'm in a good position to pass judgement. I'll have to have a better look, but seems like you were right in your prediction at least :)

Since this issue was hijacked by a new one, I suggest changing the title to help other onlookers. Or else close it and create a new one still.

**Comment by jnikula** [comment]
> I also tried to do this but I got the following issue:

Thanks, that's the same issue, just without any of the Sphinx parts, so we've excluded that. Can you share your `main.c`? Or please try to trim it down to a minimal example that you can share that produces the same issue on the command line.

**Comment by jnikula** [comment]
Just for a sanity check, does `python3 -m hawkmoth` work on any of the other files you have? Or some of the .c files from https://github.com/jnikula/hawkmoth/tree/master/test

**Comment by jnikula** [comment]
> Since this issue was hijacked by a new one, I suggest changing the title to help other onlookers. Or else close it and create a new one still.

Yeah, I really should have forked the assertion issue to a new bug from the start, but since the original issue was really something we've already fixed, I just opted to rename the bug now.

**Comment by Azzeccagarbugli** [comment]
> Just for a sanity check, does `python3 -m hawkmoth` work on any of the other files you have? Or some of the .c files from https://github.com/jnikula/hawkmoth/tree/master/test

I tried to execute the command `python3 -m hawkmoth` on the file called macro.c, as you suggested, and I got the following result:

```
.. c:macro:: FAILURE

   Failure status.


.. c:function:: DIE()

   Terminate immediately with failure status.

   See :c:macro:`FAILURE`.


.. c:function:: ARRAY_SIZE(array)

   Get the number of elements in an array.

   :param array: An array
   :return: Array size


.. c:function:: VARIADIC_MACRO(foo, ...)

   Variadic macros

   :param foo: regular argument
   :param ...: variable argument
```

So I think that the problem is in my code.

> > I also tried to do this but I got the following issue:
> 
> Thanks, that's the same issue, just without any of the Sphinx parts, so we've excluded that. Can you share your `main.c`? Or please try to trim it down to a minimal example that you can share that produces the same issue on the command line.

For sure, you can find the code at this [link](https://paste.ofcode.org/6BLBKSrDKMjCd8sZxuzPrh).

**Comment by BrunoMSantos** [comment]
Just made a very quick test and this is related to functions with an empty list of parameters.

```c
/** This works fine. */
void fun(void)
{
}

/** This does not. */
void bar()
{
}
```

I guess we just need to look for this particular case, but I haven't explored what the AST generates in detail to know what's the right™ thing to do.

**Comment by speedyleion** [comment]
It provides `TypeKind.FUNCTIONNOPROTO`
https://clang.llvm.org/doxygen/classclang_1_1FunctionNoProtoType.html#details
https://stackoverflow.com/questions/9035547/non-prototyped-function-declarations

And if you force it past the assert it returns `True` at least with clang ~9.0

**Comment by BrunoMSantos** [comment]
@speedyleion that's right.

So just for completeness, this would work fine even if it wouldn't actually compile:
```c
void foo(int a);

/** Whatever. */
void foo()
{
}
```

In the original, it's just Clang not knowing the signature of the function at that point in time and not wanting to pass any judgement on whether the function is variadic or not or whether it has any arguments at all apparently, hence 'no proto'.

In the snippet above, at the AST level, Clang hasn't even spotted the error yet: different argument number. If using `void` instead of `int a` it would have been valid C and the documentation would have worked, but the point remains.

Is there any valid non obscure C where such a function declaration would not imply a 0 argument list? I didn't test it with old style function declarations, but I think it would work fine and we shouldn't care if it doesn't.

If not, than this is a matter of potentially supporting `void foo()` as synonym for `void foo(void)`, which may be quite common actually. Supporting it trivially --- without cross checking with all other definitions of the same function in the AST --- won't catch the fact that the snippet above is invalid, but I don't think we should care about that either.

@jnikula, patch + tests sent to mailing list assuming this easy solution.

**Comment by Azzeccagarbugli** [comment]
> @speedyleion that's right.
> 
> So just for completeness, this would work fine even if it wouldn't actually compile:
> 
> ```c
> void foo(int a);
> 
> /** Whatever. */
> void foo()
> {
> }
> ```
> 
> In the original, it's just Clang not knowing the signature of the function at that point in time and not wanting to pass any judgement on whether the function is variadic or not or whether it has any arguments at all apparently, hence 'no proto'.
> 
> In the snippet above, at the AST level, Clang hasn't even spotted the error yet: different argument number. If using `void` instead of `int a` it would have been valid C and the documentation would have worked, but the point remains.
> 
> Is there any valid non obscure C where such a function declaration would not imply a 0 argument list? I didn't test it with old style function declarations, but I think it would work fine and we shouldn't care if it doesn't.
> 
> If not, than this is a matter of potentially supporting `void foo()` as synonym for `void foo(void)`, which may be quite common actually. Supporting it trivially --- without cross checking with all other definitions of the same function in the AST --- won't catch the fact that the snippet above is invalid, but I don't think we should care about that either.
> 
> @jnikula, patch + tests sent to mailing list assuming this easy solution.

So the only solution for now is waiting for your fix guys? I would like to help somehow but I have really no time recently.

Anyway I'm really glad that from that small _bug_ we found out this problematic behaviour of the software.

**Comment by jnikula** [comment]
`foo()` and `foo(void)` are **not** the same in C, the former means nothing is said about the parameters, and it's an obsolescent feature that should be avoided. If you declare a function with an empty parameter list, it's the actual function call that determines the parameters that are passed, i.e. there are no checks on the number and types of the parameters.

I presume in practically all modern code, the former is a mistake and the intent is the latter. Including @Azzeccagarbugli's example. It's a bit of a pitfall, because they **are** synonymous in C++.

Obviously Hawkmoth should not trip over this. I'm also hesitant to complain about the use of even obsolescent language features; I don't think it's our place to do so. For the sake of documentation, I think I'd just consider `foo()` and `foo(void)` the same.

**Comment by BrunoMSantos** [comment]
> foo() and foo(void) are not the same in C, the former means nothing is said about the parameters, and it's an obsolescent feature that should be avoided.

Hm, fair enough. I won't claim expertise in old C. I won't even touch the C++ case...

> I'm also hesitant to complain about the use of even obsolescent language features; I don't think it's our place to do so.

Well ok, but what would be the goal? Allowing to modernize the documentation of an old piece of code without modernizing the code? Somehow I'd put that in the very bottom of my todo list. I didn't mean we should break compatibility for the sake of breaking it of course, my point is in not bending over backwards to avoid it. And I think we'd have to do so here because we'd have to almost compile the code to get it right.

> For the sake of documentation, I think I'd just consider foo() and foo(void) the same.

+1

> So the only solution for now is waiting for your fix guys? I would like to help somehow but I have really no time recently.

If you can avoid defining functions with empty parameters, you won't need any update. In particular, change your `void printOpt()` to `void printOpt(void)`.

**Comment by jnikula** [comment]
> So the only solution for now is waiting for your fix guys?

You can move forward by making your function prototype `void printOpt(void)` i.e. by adding `void` there. I believe it's your intent anyway even if you didn't know it. ;)

@BrunoMSantos has provided the fix, which I've nitpicked a bit, but will merge otherwise as-is.

**Comment by jnikula** [comment]
> Well ok, but what would be the goal? Allowing to modernize the documentation of an old piece of code without modernizing the code? Somehow I'd put that in the very bottom of my todo list. I didn't mean we should break compatibility for the sake of breaking it of course, my point is in not bending over backwards to avoid it. And I think we'd have to do so here because we'd have to almost compile the code to get it right.

I simply meant that I don't want to add an error message or anything like that, even if we spot the case. Let's just avoid the backtrace like your patch does.

**Comment by jnikula** [comment]
Fixed by 2189bde52c99 ("parser: fix documentation of non prototyped functions").

@Azzeccagarbugli Many thanks for the report!
